engines/registration_sharing/lib/registration_sharing.rb

```diff
@@ -0,0 +1,68 @@
+$LOAD_PATH.push File.expand_path(__dir__, '..')
+
+require 'registration_sharing/engine'
+
+module RegistrationSharing
+  RMT_REGSHARING_DEFAULT_DATA_DIR = '/var/lib/rmt/regsharing'.freeze
+
+  class << self
+    # creates a file on the disk to mark systems with changed registration data
+    def save_for_sharing(obj)
+      peers = config_peers
+      return unless peers
+
+      system = (obj.class <= ::System) ? obj : obj.system
+
+      peers.each do |peer|
+        write_file(peer, system.login)
+      end
+    end
+
+    def config_peers
+      peers = Settings[:regsharing][:peers] rescue nil
+      return if peers.blank?
+
+      (peers.class == String) ? [peers] : peers
+    end
+
+    def config_data_dir
+      Settings[:regsharing][:data_dir] || RMT_REGSHARING_DEFAULT_DATA_DIR
+    rescue StandardError
+      RMT_REGSHARING_DEFAULT_DATA_DIR
+    end
+
+    def config_api_secret
+      Settings[:regsharing][:api_secret] rescue nil
+    end
+
+    def config_ca_path
+      Settings[:regsharing][:ca_path] rescue nil
+    end
+
+    protected
+
+    def write_file(peer, system_login)
+      dirname = File.join(config_data_dir, peer)
+      FileUtils.mkpath(dirname)
+
+      filename = File.join(config_data_dir, peer, system_login)
+
```

Comment:
Is system_login unique for every invocation of write_file? If not we have the same dilemma with flock here as discussed w.r.t. SMT.

Reply:
`write_file` dumps the timestamp (with microseconds) into the file. After the registration is shared, the process would lock the file and read it's contents -- if the file was modified during regsharing call, it will not be unlinked and will be synced again in the next iteration. If the file wasn't modified, it is removed.

Reply:
`system_login` is the login of the client system that had it's registration data modified. So yes, it will create a unique file for each individual system.

It is possible for the system to change while it's being synced (an additional product could be activated). That's why the file isn't locked when API requests are being made (so that the web workers don't block in `flock` call). If the file has been modified during sync, it won't be deleted and will be re-synced in the next iteration.

Reply:
OK